targets/TARGET_NORDIC/TARGET_NRF5x/flash_api.c

```diff
@@ -38,82 +38,167 @@
+         * SoftDevice API should work both when the SoftDevice is enabled or disabled.
+         * NVMC API is used when the SoftDevice is not present.
+         */
 #if defined(SOFTDEVICE_PRESENT)
-    uint8_t sd_enabled;
-    if ((sd_softdevice_is_enabled(&sd_enabled) == NRF_SUCCESS) && sd_enabled == 1) {
-        return -1;
-    }
+        result = nrf_fstorage_init(&nordic_fstorage, &nrf_fstorage_sd, NULL);
+#else
+        result = nrf_fstorage_init(&nordic_fstorage, &nrf_fstorage_nvmc, NULL);
 #endif
-    return 0;
+
+        /* Initialize low power ticker for timeouts. */
+        lp_ticker_init();
+    }
+
+    /* Convert Nordic SDK error code to mbed HAL. */
+    return (result == NRF_SUCCESS) ? 0 : -1;
 }
 
 int32_t flash_free(flash_t *obj)
 {
     (void)(obj);
+
     return 0;
 }
 
 int32_t flash_erase_sector(flash_t *obj, uint32_t address)
 {
     (void)(obj);
-#if defined(SOFTDEVICE_PRESENT)
-    uint8_t sd_enabled;
-    if ((sd_softdevice_is_enabled(&sd_enabled) == NRF_SUCCESS) && sd_enabled == 1) {
-        return -1;
+
+    ret_code_t result = NRF_ERROR_NO_MEM;
+
+    /* Setup stop watch for timeout. */
+    uint32_t start_us = lp_ticker_read();
+    uint32_t now_us = start_us;
+
+    /* Retry if flash is busy until timeout is reached. */
+    while (((now_us - start_us) < PAGE_ERASE_TIMEOUT_US) &&
+           (result == NRF_ERROR_NO_MEM)) {
+
+        /* Map mbed HAL call to Nordic fstorage call. */
+        result = nrf_fstorage_erase(&nordic_fstorage, address, 1, NULL);
+
+        /* Read timeout timer. */
+        now_us = lp_ticker_read();
     }
-#endif
-    nrf_nvmc_page_erase(address);
-    return 0;
+
+    /* Convert Nordic SDK error code to mbed HAL. */
+    return (result == NRF_SUCCESS) ? 0 : -1;
 }
 
 int32_t flash_program_page(flash_t *obj, uint32_t address, const uint8_t *data, uint32_t size)
 {
-#if defined(SOFTDEVICE_PRESENT)
-    uint8_t sd_enabled;
-    if ((sd_softdevice_is_enabled(&sd_enabled) == NRF_SUCCESS) && sd_enabled == 1) {
-        return -1;
+    (void)(obj);
+
+    /* Check that data pointer is not NULL. */
+    ret_code_t result = NRF_ERROR_NULL;
+
+    if (data) {
+
+        /* nrf_fstorage_write only accepts word aligned input buffers.
+         * Cast data pointer to word pointer and read word into temporary word variable
+         * which should be word aligned. This should be safe on Cortex-m4 which is able
+         * to load unaligned data.
+         */
+        const uint32_t *data_word = (const uint32_t *) data;
+
+        /* Loop through input buffer 4 bytes at a time. */
+        for (size_t index = 0; index < (size / 4); index++) {
+
+            /* Load 4 bytes into temporary variable. */
+            uint32_t word = data_word[index];
+
+            /* Setup stop watch for timeout. */
+            uint32_t start_us = lp_ticker_read();
+            uint32_t now_us = start_us;
+
+            /* Retry if flash is busy until timeout is reached. */
+            do {
```

Comment:
I suggest using the same while loop as in erase API , why doing it different?

Reply:
For the second iteration through the for-loop, result will be NRF_SUCCESS and not NRF_ERROR_NO_MEM, so a while-loop would fail the first iteration unless we explicitly reset result to NRF_ERROR_NO_MEM. 

By making it into a do-while-loop, result is automatically on the first call to nrf_fstorage_write.